src/Bicep.Core/TypeSystem/Az/AzTypeLoaderOci.cs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.IO;
+using Azure.Bicep.Types;
+using System.Formats.Tar;
+using System.IO.Compression;
+using System;
+using System.Collections.Generic;
+
+namespace Bicep.Core.TypeSystem.Az
+{
+    public class AzTypeLoaderOci : TypeLoader
+    {
+        private readonly Dictionary<string, byte[]> typesCache = new(StringComparer.OrdinalIgnoreCase);
+
+        public AzTypeLoaderOci(string pathToGzip) : base()
+        {
+            using var fileStream = File.OpenRead(pathToGzip);
+            using var gzipStream = new GZipStream(fileStream, CompressionMode.Decompress);
+            using var tarReader = new TarReader(gzipStream, leaveOpen: true);
+            while (tarReader.GetNextEntry(copyData: true) is { } entry)
+            {
+                if (entry.DataStream is null)
+                {
+                    throw new ArgumentException($"Failed to restore {entry.Name} from OCI provider data", nameof(entry.Name));
+
+                }
+                using var ms = new MemoryStream();
+                entry.DataStream.CopyTo(ms);
+                typesCache.Add(entry.Name, ms.ToArray());
+            }
+        }
```

Comment:
I'd suggest moving the initializatin code to a static `Create` method and keep the constructor simple as a best practice. Then we can change the `typesCache` dictionary to an `ImmutableDictionary`. For example:

```C#
private readonly ImmutableDictionary<string, byte[]> typesCache;

public OciAzTypeLoader(ImmutableDictionary<string, byte[]> typesCache)
{
    this.typesCache = typesCache;
}

public static OciAzTypeLoader Create()
{
    var typesCache = ImmutableDictionary.CreateBuilder<string, byte[]>();

    // ... initialization of typesCache

    return new(typesCache);
}
```

Reply:
Good suggestion, can you share the reason why this is desirable?

Reply:
For constructors, the principle is that a constructor should do as little as possible. Complex operations or heavy computations are better placed in separate methods or classes to keep the constructor focused and maintainable. This also avoids the need for handling exceptions in the constructor.

The use of `ImmutableDictionary` is to avoid unexpected changes to the dictionary after initialization.

Reply:
thank you for the clear explanation, TIL